docs/functionaldatatypes/zset.md

```diff
@@ -0,0 +1,276 @@
+
+This requires a `Commutative` defined on the `Sum` of the measure type to tell us how to do this addition.
+
+```scala mdoc
+val janeShoppingList: ZSet[String, Int] =
+  ZSet.fromMap(Map("apples" -> 4, "bananas" -> 5))
+
+val johnShoppingList: ZSet[String, Int] =
+  ZSet.fromMap(Map("eggs" -> 12, "bananas" -> 1))
+
+val combinedList: ZSet[String, Int] =
+  janeShoppingList.combine(johnShoppingList)
+```
+
+The result is `ZSet("apples" -> 4, "bananas" -> 6, "eggs" -> 12)`, the combination of what each of them wants to buy.
+
+## Set Product
+
+We can also define the product of two sets.
+
+This is the Cartesian product of the two sets, with each combination of elements from the two sets appearing in the product set and the measure of how many times an element appears being the product of the two measures. This requires a `Commutative` defined on the `Prod` of the measure type to tell us how to do this multiplication.
+
+A nice example of this is modeling discrete probabilities. We can define the set of all possible outcomes of a dice roll along with their associated probabilities like this:
+
+```scala mdoc
+val die: ZSet[Int, Double] =
+  ZSet.fromMap(
+    Map(
+      1 -> 1.0 / 6.0,
+      2 -> 1.0 / 6.0,
+      3 -> 1.0 / 6.0,
+      4 -> 1.0 / 6.0,
+      5 -> 1.0 / 6.0,
+      6 -> 1.0 / 6.0
+    )
+  )
+```
+
+The product of this set with itself represents the set of all possible outcomes of rolling two dice.
+
+```scala mdoc
+val pair: ZSet[(Int, Int), Double] =
+  die.zip(die)
+```
+
+This set has thirty six elements, corresponding to each possible combination of outcomes from the two dice rolls, each with a probability of one in thirty six.
+
+This gets more interesting if we say that matters is not the individual dice rolls but their sum.
+
+We can express this in code with the `zipWith` operator, which allows us to combine the elements from the product of the two sets with a function. If multiple elements from the product are mapped to the same value the measures of how many times each value appears will be added, using a `Commutative` instance defined on the `Sum` of the measure type.
+
+```scala mdoc
+val sum: ZSet[Int, Double] =
+  die.zipWith(die)(_ + _)
+```
+
+The resulting set has eleven elements, corresponding to the possible sums from two dice rolls of two to twelve. The measure for each value corresponds to the probability of that sum being rolled.
+
+So for example the probability of seven is about 16.7% whereas the probability of rolling two or twelve is only about 2.8%.
+
+We can take this even further using the `flatMap`.
+
+The `zipWith` operator lets us combine two sets when the sets are independent of each other. With the `flatMap` operator we can create a new set based on each element of the first set and then "flatten" the resulting set down.
+
+We could rewrite the example above using `flatMap` and a _for comprehension_ like this:
```

Comment:
nitpick: I don't like the style of `for comprehension` here.

Reply:
Changed.